utils_tf2.py

```diff
@@ -0,0 +1,170 @@
+import tensorflow as tf
+import numpy as np
+import torch
+
+class ModelAdapter():
+    def __init__(self, model, num_classes=10):
+        """
+        Please note that model should be tf.keras model without activation function 'softmax'
+        """
+        self.num_classes = num_classes
+        self.tf_model = model
+        self.__check_channel_ordering()
+
+    def __check_channel_ordering(self):
+        
+        for L in self.tf_model.layers:
+            if isinstance(L, tf.keras.layers.Conv2D):
+                print("[INFO] set data_format = '{:s}'".format(L.data_format))
+                self.data_format = L.data_format
+                return
+
+        print("[INFO] Can not find Conv2D layer")
+        input_shape = self.tf_model.input_shape
+
+        if input_shape[3] == 3:
+            print("[INFO] Because detecting input_shape[3] == 3, set data_format = 'channels_last'")
+            self.data_format = 'channels_last'
+
+        elif input_shape[3] == 1:
+            print("[INFO] Because detecting input_shape[3] == 1, set data_format = 'channels_last'")
+            self.data_format = 'channels_last'
+
+        else:
+            print("[INFO] set data_format = 'channels_first'")
+            self.data_format = 'channels_first'
+
+    def _get_logits(self, x_input):
+        logits = self.tf_model(x_input, training=False)
+        return logits
+
+    def _get_jacobian(self, x_input):
+        with tf.GradientTape(watch_accessed_variables=False) as g:
+            g.watch(x_input)
+            logits = self._get_logits(x_input)
+
+        jacobian = g.batch_jacobian(logits, x_input)
+
+        if self.data_format == 'channels_last':
+            jacobian = tf.transpose(jacobian, perm=[0,1,4,2,3])
+
+        return jacobian
+
+    def _get_xent(self, logits, y_input):
+        xent   = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits, labels=y_input)
+        return xent
+
+    def _get_grad_xent(self, x_input, y_input):
+        with tf.GradientTape(watch_accessed_variables=False) as g:
+            g.watch(x_input)
+            logits = self._get_logits(x_input)
+            xent = self._get_xent(logits, y_input)
+        
+        grad_xent = g.gradient(xent, x_input)
+
+        return logits, xent, grad_xent
+
+    def _get_dir(self, logits, y_input):
```

Comment:
`_get_dir` should be renamed `_get_dlr` (and the same below)

Reply:
OK, I will fix format latter.